### ISSUE TITLE
Determine when cases are ready for distribution

### ISSUE THREAD

**Issue by amprokop** [department-of-veterans-affairs/caseflow#8288]
### Hearing docket rules
- has the hearing occurred, and has it been transcribed - OR - has it been withdrawn AND has the date at which it would have been surfaced to hearing staff for scheduling passed
- has the 90 day window for evidence submission expired or been waived
- if there is a VSO and the VSO writes IHP for hearings docket cases (PVA is the only one right now), has the 45? day window for IHP writing passed or has it been marked complete

### Evidence Submission docket rules
- has the 90 day window for evidence submission expired or been waived
- if there is a VSO, has the 45? day window for IHP writing passed or has it been marked complete

### Direct docket rules
- if there is a VSO, has the 45? day window for IHP writing passed or has it been marked complete

## Implementation Option 1 - DistributionTasks

We could decouple the process of looking for cases that are ready for distribution from the auto-case distribution action itself.

When a judge clicks "Request Cases", we would simply find all appeals that have incomplete DistributionTasks. 

### Direct docket
  - If the appeal has no VSO, create a DistributionTask at intake
  - otherwise, in a nightly job, look for direct docket appeals with no DistributionTask but with an InformalHearingPresentationTask that is either incomplete and more than 45 days old or complete.
 
### Evidence Submission
  - In a nightly job, look for evidence submission docket appeals where the 90 day window has expired, or a waiver has been submitted (will need an `EvidenceSubmissionWindowWaiver` class). If the appeal does not have a VSO, create a DistributionTask. Otherwise, create an IHP task
- In a nightly job, look for evidence submission appeals with no DistributionTask but with an InformalHearingPresentationTask that is either incomplete and more than 45 days old or complete

### Hearing docket
- In a nightly job, look for cases where the hearing has occurred, and has it been transcribed - OR - has it been withdrawn AND has the date at which it would have been surfaced to hearing staff for scheduling passed. If so, create an EvidenceSubmissionTask.
- In a nightly job, look for hearing docket appeals where the 90 day window has expired, or a waiver has been submitted (will need an `EvidenceSubmissionWindowWaiver` class). If the appeal does not have a VSO that writes IHPs (PVA is the only one we know of that writes IHPs for hearing docket cases), create a DistributionTask. Otherwise, create an IHP task.
- In a nightly job, look for hearing docket appeals with no DistributionTask but with an InformalHearingPresentationTask that is either incomplete and more than 45 days old or complete.

## Implementation Option 2 — query when users request cases, no DistributionTask

Leaning away from this option since it may be a very complicated query that may be quite slow.

**Comment by laurjpeterson** [comment]
I think having a DistributionTask makes sense - seems nice and simple!

Typically, tasks have helped us assign responsibility to a case. I'm assuming this task would not need to be assigned to a particular team/organization or person, correct?

 I also think we'd want to hide this task from Case Timeline. 

To know whether the 90-day evidence window has passed or has been waived, I was thinking we could create a 90-day evidence window task (see: https://github.com/department-of-veterans-affairs/caseflow/issues/7916). If an appellant waives their 90-day evidence window, either in a hearing or by sending in correspondence, a BVA user could "cancel" or "close" the 90-day evidence window task. Could this prevent the need for an evidence submission waiver class?

**Comment by lowellrex** [comment]
I don't think option 2 will result in a complicated or slow-running query (should be querying a single table, `tasks` with several conditions in the where clause), but I lean towards option 1 since I think it is a pretty tidy solution and nests well with the pattern we've already established with tasks.

As for how we implement this, I think we can make use of [`RootTask.when_child_task_completed`](https://github.com/department-of-veterans-affairs/caseflow/blob/a68fc23b5124b127837f6116d2782d1a45924b74/app/models/tasks/root_task.rb#L8) to check if this case has been distributed (if the task tree has a distribution task or not, though it might need to be more complex than that to account for all of the cases that do not yet have distribution tasks). This way we could offload the triggering of distribution task creation onto the child task sending a message (that the child task has been completed) to the root task which then runs some logic of its own.

One additional feature that may make this easier (and that @mdbenjam and I have been discussing for a while now), is the introduction of a timed task that could be inherited by IHP and evidence submission tasks (and be created as children of ColocatedTasks that we want to put on hold, the use case that Mark and I had been considering most proximately for this timed task). The job that would then run every hour (or day or minute) would be responsible for closing out expired timed tasks instead of doing work on tasks that is precisely related to the case distribution process even though we could use it for that purpose. It's really six of one and a half dozen of the other (timed tasks vs a specific to close case distribution-related tasks), but perhaps the timed task functionality has more broad applications that we can leverage elsewhere.

**Comment by cmgiven** [comment]
I also like the simplicity of option 1. I also +1 Lowell's suggestion of using `when_child_task_completed` so as to avoid nightly jobs. The only problem may be that I believe we don't intend to complete IHP tasks when they expire. So that might still require a nightly job to check for expired IHP tasks.

The existence of a mail task or other colocated task would also block distribution. So we might need something parallel to `when_child_task_completed` called `when_child_task_created` to potentially delete the DistributionTask if the case is no longer eligible for distribution.

**Comment by sharonwarner** [comment]
I've only thought through the hearings docket, but I'm wondering if, instead of a nightly job, we could create the DistributionTask when the hearing is held or withdrawn, but then put various on hold tasks underneath the DistributionTask (Transcription & EvidenceWindow).

**Comment by lowellrex** [comment]
I really like the idea of a `when_child_task_created` method @cmgiven! We currently update the parent task's status to `on_hold` when a child task is created and it would be nice to move that to `when_child_task_created` so that the parent task can own that behaviour itself.

**Comment by amprokop** [comment]
Re: `when_child_task_completed` My current thought is — if we're going to need a nightly job for some distribution task creation (evidence period and IHP period expiration), I'd like to put all distribution task creation in the job, both to keep all the logic in one place for readability, and so we can keep failure, monitoring, and error handling in one place. 

@cmgiven @lowellrex — how would you feel about having all the distribution task logic start off in a job, then if we get timed tasks, consider distributing it to various hooks?

**Comment by lowellrex** [comment]
@amprokop I'm wary of cronjobs because it is not a pattern we use all that frequently and therefore don't have much experience with them as a team and haven't common practices about how or when to implement them.

I think we can create and update `DistributionTask`s correctly by hooking into the various lifecycle events of tasks as is and accomplish objective without needing any tasks at all. The only exception would be expiring IHP or evidence submission tasks, but I think those should have their own mechanism for completing that is very specific to those tasks (and [maybe all "timed tasks"](https://github.com/department-of-veterans-affairs/caseflow/issues/8288#issuecomment-447314137)), but once that job is in place then we will have all the hooks we need for the creation and modification of `DistributionTask`s.

**Comment by cmgiven** [comment]
@amprokop How would you address an appeal no longer being ready for distribution because of a new mail task, in the job model?

**Comment by mdbenjam** [comment]
Some thoughts around this.

I generally like the idea of a distribution task. I agree with @lowellrex that we should have a simple cron job which closes timed tasks, and then let the logic of creating new tasks live within each task type in `when_child_task_completed`. I also think @sharonwarner has suggested a pretty powerful way to address the problem @cmgiven just brought up. If we create the distribution task early (say at intake), we can add blocking tasks to it. Sharon suggested doing this with transcription and evidence window, but I think it applies equally well to VSO IHP tasks and mail tasks.

Addressing one of @cmgiven's concerns. 

> The only problem may be that I believe we don't intend to complete IHP tasks when they expire. So that might still require a nightly job to check for expired IHP tasks.

I wonder if we should make two types of IHP tasks. `IHPTask` and `PastDueIHPTask`. The former blocks distribution. The second doesn't. i.e. the first is a child of the `DistributionTask`, the latter is a child of the `RootTask`.

This also fits in well with what @laurjpeterson suggested:

> If an appellant waives their 90-day evidence window, either in a hearing or by sending in correspondence, a BVA user could "cancel" or "close" the 90-day evidence window task. Could this prevent the need for an evidence submission waiver class?

If the evidence window is a child of the distribution task, then when we close or cancel it, the case is automatically ready for distribution.

**Comment by amprokop** [comment]
> @amprokop I'm wary of cronjobs because it is not a pattern we use all that frequently and therefore don't have much experience with them as a team and haven't common practices about how or when to implement them.

@lowellrex — hmm, I think maybe you might just be missing context on this. we have a pretty wide variety of nightly/more frequent scheduled jobs that run for Dispatch (preparing dispatch tasks), Intake (syncing various items with VBMS), Reader (fetching documents), as well as checking dependencies, reminders if apps are out of service, and others.

**Comment by amprokop** [comment]
```
@amprokop How would you address an appeal no longer being ready for distribution because of a new mail task, in the job model?
```

I think maybe @mdbenjam's answer is the best. When making the mail task, we put the DistributionTask on hold or create it as a child task.

**Comment by amprokop** [comment]
```
I wonder if we should make two types of IHP tasks. IHPTask and PastDueIHPTask. The former blocks distribution. The second doesn't. i.e. the first is a child of the DistributionTask, the latter is a child of the RootTask.
```

@mdbenjam — maybe you could explain why `PastDueIHPTask` makes sense as a task. I've noticed your suggestions in other threads that we should treat more things as tasks that aren't necessarily an actionable item for a person to complete (e.g. `VsoVisibilityTask`). Wondering how you would respond to the pushback that we're overloading the concept of a task.

**Comment by amprokop** [comment]
Overall, though, I like the feedback that y'all are giving around creating children tasks, and think that approach is generally more consistent.

So for example:

-at Intake *always* create a DistributionTask.
-for all VSO-having IHP-writing appeals, create an open IHP task and put the DistributionTask on hold.
-for all evidence submission appeals, create an `EvidenceSubmissionTask` and put the IHP task on hold.- -for hearing appeals, create some hearing task that blocks EvidenceSubmission and IHP.
-if a mail task comes in, create it as a child on the DistributionTask and ensure it blocks the Distribution task being set to complete

**Comment by amprokop** [comment]
@cmgiven — bah, humbug. Interesting thought. Some items to think through here:

For direct docket appeals without an IHP-writing POA at first, they are initially ready for distribution, but if they receive an IHP-writing POA at some point before they are distributed to the judge, do they then need to be pulled out of being ready for distribution and given an IHP-writing period? What about if they receive an IHP-writing POA *after* the case is assigned to the judge?

How do we even detect that POA has changed after intake and before distribution? Right before distribution, do we call BGS to check if POA has changed? If we do that, auto case distribution is at risk of becoming very slow and unreliable. Alternatively, this happens as part of a nightly job, where we check for POA changes for the first N appeals that will be distributed.

**Comment by amprokop** [comment]
> If an appellant waives their 90-day evidence window, either in a hearing or by sending in correspondence, a BVA user could "cancel" or "close" the 90-day evidence window task. Could this prevent the need for an evidence submission waiver class?
> 
> If the evidence window is a child of the distribution task, then when we close or cancel it, the case is automatically ready for distribution.

I like this idea. Even if there is already a child task, we could cancel the evidence submission task. And we could have a helper method `waived?` that just checks if the task is cancelled.

**Comment by mdbenjam** [comment]
AFAIK we haven't implemented a `canceled` state. We wanted more use cases before we did. Not sure if this pushes that over the edge.

Regarding changing POA or new POA. I think we need a nightly job to check for this, but I think that's a separate discussion. In fact, we're having a discussion around VSO visibility tasks later today where I think we'll have to consider this.

> @mdbenjam — maybe you could explain why PastDueIHPTask makes sense as a task. I've noticed your suggestions in other threads that we should treat more things as tasks that aren't necessarily an actionable item to be completed (e.g. VsoVisibilityTask). Wondering how you would respond to the pushback that we're overloading the concept of a task.

My basic answer is, creating subclasses of Task is easy. We should do that where we would otherwise need to create some logic for the original case that doesn't fit into the task model. For example, let's say we just have an `IHPTask` and create a new type of `TimerTask` (which @lowellrex  and I have discussed.) A timer task would last a certain number of days and then expire, taking its parent task off hold. If the `IHPTask` is a child of our `DistributeTask` then we have to have some pretty complex logic when the `IHPTask`'s child `TimerTask` expires. We would likely need to move it to have the `RootTask` as its parent instead of the `DistributeTask`. In my view that behavior breaks the task model. Instead being immutable and creating a new task whose parent is the `RootTask` feels cleaner. This could still be an `IHPTask`. The only reason I'd suggest a new task type is so that we could easily display it in the UI. Also there might be other actions we need to take when someone completes an `IHPTask` before the deadline vs after (though this is speculative).

**Comment by amprokop** [comment]
@mdbenjam — I'm starting to see what you mean, I 

> If the IHPTask is a child of our DistributeTask then we have to have some pretty complex logic when the IHPTask's child TimerTask expires. 

Can you help me understand what the complexity is here? Don't we just complete the IHPTask, which then triggers a `when_child_task_complete` or something hook on DistributionTask, which may or may not take it off of hold depending on whether it has any other open child tasks?

> We would likely need to move it to have the RootTask as its parent instead of the DistributeTask
🤔 — can you explain why this would be the case?

**Comment by sharonwarner** [comment]
For hearings schedule, we're going to use ScheduleHearingTasks to determine which AMA appeals need to be scheduled for a hearing. Before case distribution, the appeal will also need a completed TranscriptionTask and a completed EvidenceWindowTask. I think it makes sense to have the TranscriptionTask and the EvidenceWindowTask live as on hold tasks underneath a CaseDistributionTask. So the path would be something like Intake -> ScheduleHearingTask -> HoldHearingTask -> CaseDistributionTask (TranscriptionTask/EvidenceWindowTask).

**Comment by mdbenjam** [comment]
@amprokop we'd need to move it to the root task because we (I think) still want it to be visible to VSOs to write an IHP, even if they didn't meet the deadline. If we don't want it to block auto case distribution then we'd have to move it to be a child of `RootTask`.

@sharonwarner I think that mostly makes sense, though I don't know that `CaseDistributionTask` should be under `HoldHearingTask`? Don't we want to hold the hearing before distributing it to a judge for writing the decision? Might be wrong on this.

**Comment by amprokop** [comment]
> @amprokop we'd need to move it to the root task because we (I think) still want it to be visible to VSOs to write an IHP, even if they didn't meet the deadline.

Interesting... I did not know this was a need, thank you for clarifying. Yeah in that case you're right — it would make sense to have two tasks, and open the second one on the RootTask only when the first one expires (would the past due task type have to be closed eventually as well? I'm not sure we'd want VSOs to be writing IHPs for cases that already have decisions, for example).

**Comment by amprokop** [comment]
https://github.com/department-of-veterans-affairs/caseflow/issues/8336 opened this to track POA changing and what to do there

**Comment by mdbenjam** [comment]
I could be wrong on that requirement, but I think that's how IHPs have been treated in the past. Actually I think in legacy world they block moving to write a decision forever, but from my understanding we want to still provide folks an opportunity to write IHPs even if the window has closed. @laurjpeterson to confirm.

**Comment by amprokop** [comment]
Created a list of discrete tasks related to this here
https://github.com/department-of-veterans-affairs/caseflow/issues/8338

**Comment by laurjpeterson** [comment]
Emailed BVA SMEs about which specific mail tasks should block distribution to judges today.

**Comment by laurjpeterson** [comment]
@amprokop here you go!

**Mail types that should prevent case distribution to judges**
- Congressional Inquiries
- FOIA request
- Privacy act request
- Privacy complaint (to be confirmed with Meg/Jessie)
- Power of attorney related
- Extension request
- Hearing related

**Rest of mail types, mail types that do not prevent case distribution**
- Status Inquiry
- CUE-related
- Change of address
- Controlled correspondence
- Death certificate
- Evidence or argument
- Motion for reconsideration
- Motion to Advance on Docket
- Motion to vacate
- Other motion
- Returned or undeliverable mail (did not move above, because it only potentially blocks distribution)
- Withdrawal of appeal

**Comment by laurjpeterson** [comment]
Closing, because this has been implemented
